pkg/sentry/socket/netlink/route/protocol.go

```diff
@@ -500,6 +500,79 @@ func (p *Protocol) delAddr(ctx context.Context, msg *netlink.Message, ms *netlin
 	return nil
 }
 
+func fetchRoute(msg *netlink.Message) (*inet.Route, *syserr.Error) {
+	var rtMsg linux.RouteMessage
+	attrs, ok := msg.GetData(&rtMsg)
+	if !ok {
+		return nil, syserr.ErrInvalidArgument
+	}
+	if rtMsg.Table != linux.RT_TABLE_MAIN {
+		// Only the main table since we don't have multiple
+		// routing tables.
+		return nil, syserr.ErrInvalidArgument
+	}
+	route := inet.Route{
+		Family:   rtMsg.Family,
+		DstLen:   rtMsg.DstLen,
+		SrcLen:   rtMsg.SrcLen,
+		TOS:      rtMsg.TOS,
+		Table:    rtMsg.Table,
+		Protocol: rtMsg.Protocol,
+		Scope:    rtMsg.Scope,
+		Type:     rtMsg.Type,
+		Flags:    rtMsg.Flags,
+	}
+
+	attHdr, buff, rest, ok := attrs.ParseFirst()
+	for ok {
+		switch attHdr.Type {
+		case linux.RTA_DST:
+			route.DstAddr = buff
+		case linux.RTA_SRC:
+			route.SrcAddr = buff
+		case linux.RTA_GATEWAY:
+			route.GatewayAddr = buff
+		case linux.RTA_OIF:
+			// This conversion is also present in the kernel...
+			var outputIF primitive.Int32
+			outputIF.UnmarshalUnsafe(buff)
+			route.OutputInterface = int32(outputIF)
+		default:
+			// Other attributes are not implemented yet.
+			return nil, syserr.ErrInvalidArgument
+		}
+		attHdr, buff, rest, ok = rest.ParseFirst()
+	}
+	return &route, nil
+}
+
+// newRoute handles RTM_NEWROUTE requests.
+func (p *Protocol) newRoute(ctx context.Context, msg *netlink.Message, ms *netlink.MessageSet) *syserr.Error {
```

Comment:
We'll definitely need to have some tests. See: test/syscalls/linux/socket_netlink_route.cc
https://cs.opensource.google/gvisor/gvisor/+/master:test/syscalls/linux/socket_netlink_route.cc

Reply:
+1. Please add tests for the following

a) NetLink AddRoute command
b) Stack.AddRoute/SetRouteTable in stack_test.go and verify that routes are sorted in the correct expected order.
c) Stack.FindRoute() to ensure that the correct route is picked when multiple matching routes exist in decreasing order of specificity.